tmt/export/__init__.py

```diff
@@ -0,0 +1,424 @@
+# Until Bugzilla gets its own annotations and recognizable imports...
+BugzillaInstance = Any
+
+
+# TODO: why this exists?
+log = fmf.utils.Logging('tmt').logger
+
+
+# For linking bugs
+BUGZILLA_XMLRPC_URL = "https://bugzilla.redhat.com/xmlrpc.cgi"
+RE_BUGZILLA_URL = r'bugzilla.redhat.com/show_bug.cgi\?id=(\d+)'
+
+
+# ignore[type-arg]: bound type vars cannot be generic, and it would create a loop anyway.
+ExportableT = TypeVar('ExportableT', bound='Exportable')  # type: ignore[type-arg]
+ExportClass = Type['ExportPlugin']
+
+_RawExportedInstance = Dict[str, Any]
+_RawExportedCollection = List[_RawExportedInstance]
+_RawExported = Union[_RawExportedInstance, _RawExportedCollection]
+
+
+# Protocols describing export methods.
+class FmfIdExporter(Protocol):
+    def __call__(self, fmf_id: 'tmt.base.FmfId') -> str:
+        pass
+
+
+class InstanceExporter(Protocol):
+    def __call__(self, instance: ExportableT, keys: Optional[List[str]] = None) -> str:
+        pass
+
+
+class CollectionExporter(Protocol):
+    def __call__(self, collection: List[ExportableT], keys: Optional[List[str]] = None) -> str:
+        pass
+
+
+class Exportable(Generic[ExportableT]):
+    """ Mixin class adding support for exportability of class instances """
+
+    # Declare export plugin registry as a class variable, but do not initialize it. If initialized
+    # here, the mapping would be shared by all classes, which is not a desirable attribute.
+    # Instead, mapping will be created by provides_*_export decorators.
+    _export_plugin_registry: ClassVar[Dict[str, ExportClass]]
+
+    # Keep this method around, to correctly support Python's method resolution order.
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        super().__init__(*args, **kwargs)
+
+    # Cannot use @property as this must remain classmethod
+    @classmethod
+    def get_export_plugin_registry(cls) -> Dict[str, ExportClass]:
+        if not hasattr(cls, '_export_plugin_registry'):
+            cls._export_plugin_registry = {}
+
+        return cls._export_plugin_registry
+
+    @classmethod
+    def provides_export(cls, format: str) -> Callable[[ExportClass], ExportClass]:
+        """
+        A decorator for registering export format.
+
+        Decorate an export plugin class to register a format.
+        """
+
+        def _provides_export(export_cls: ExportClass) -> ExportClass:
+            cls.get_export_plugin_registry()[format] = export_cls
+
+            return export_cls
+
+        return _provides_export
+
+    @classmethod
+    def _get_collection_exporter(cls, format: str) -> CollectionExporter:
+        exporter_class = cls.get_export_plugin_registry().get(format, None)
+
+        if exporter_class is None:
+            raise tmt.utils.GeneralError(
+                f"Export format '{format}' not supported for {cls.__name__.lower()}.")
+
+        return cast(CollectionExporter, getattr(
+            exporter_class, f'export_{cls.__name__.lower()}_collection'))
```

Comment:
So if we removed the `_get_instance_exporter` in d35bc357e67ef7e67e531d005e4678581898a4cb the `_collection` suffix probably can be removed?

Reply:
Dropped in 2161623473fc3812fcc15c9fbbfe3c69afa9c235.

Note that I left the `collection` substring in export plugin methods, e.g. `export_story_collection`, they do export a collection after all.

Reply:
I understood that the suffix was necessary to differentiate from the instance exporting. Seems a bit unnecessary now when we to only collection exporting. But I can live with that.

Reply:
There are also `{Test,Plan,Story}.{export,export_collection}` methods, where `export` is a wrapper for `export_collection`, therefore keeping `collection` around seems like a better choice, to not be mistaken for `Tmt.export`, for example.

But I do understand, we could get away with `ExportPlugin.export_tests`, although the `collection` bit aligns nicely for me in the chain of `Test.export_collection` => `YAMLExporter.export_test_collection`.